apollo-gradle-plugin-kotlin/src/main/kotlin/com/apollographql/apollo/gradle/kotlin/ApolloPlugin.kt

```diff
@@ -0,0 +1,243 @@
+package com.apollographql.apollo.gradle.kotlin
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.internal.tasks.factory.dependsOn
+import com.apollographql.apollo.compiler.GraphQLCompiler
+import org.gradle.api.Action
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.internal.HasConvention
+import org.gradle.api.internal.file.FileResolver
+import java.io.File
+import org.gradle.api.plugins.JavaPluginConvention
+import org.gradle.api.tasks.TaskProvider
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+import javax.inject.Inject
+import kotlin.IllegalArgumentException
+
+
+class ApolloPlugin : Plugin<Project> {
+  companion object {
+    const val TASK_GROUP = "apollo"
+  }
+
+  lateinit var apolloExtension: ApolloExtension
+
+  override fun apply(project: Project) {
+    apolloExtension = project.extensions.create("apollo", ApolloExtension::class.java)
+
+    // the extension block has not been evaluated yet, register a callback once the project has been evaluated
+    project.afterEvaluate {
```

Comment:
That's was the main reason why extension was written on grovvy with magic of `Property` that will be evaluated later during execution, so they are placeholders for values. Is there any way we can do this in Kotlin?

Reply:
We can certainly do the same in Kotlin but that wouldn't remove the race between:

1. The Android variants or java/kotlin sourcesets are available
2. The buildscript is evaluated and the extension populated.

I believe it only works with the current plugin because the android plugin itself calls `afterEvaluate`

Reply:
I can't find anymore the `afterEvaluate` I thought I had seen in the Android plugin...
Still, removing the `afterEvaluate` makes it implicit that the extension is populated when the tasks are configured.

Reply:
To build on my previous answer, we could move the validation of the apollo extension to the execution phase (when taskAction) is called and used `Property` everywhere. That should work since the extension is certainly populated by the time execution starts.

The only thing left would be the `generateKotlinModels` field that I believe is required earlier, to wire the `compileJava/compileKotlin` to the `codegenTasks` ([link](https://github.com/martinbonnin/apollo-android/blob/53e52ebfd0c8a74eb20af9b5addc7b4636c2db69/apollo-gradle-plugin-incubating/src/main/kotlin/com/apollographql/apollo/gradle/kotlin/ApolloPlugin.kt#L72))

That I'm not 100% sire how to do yet...

Reply:
We can try to register both tasks for Java and Kotlin, and skip execution depending on the flag `generateKotlinModels ` and if Kotlin plugin is applied?

Reply:
The issue is if we add an empty file set to the javaCompile input it'll break (see https://github.com/apollographql/apollo-android/pull/1353). Or can we tell everyone to skip the `compileJava` based on `generateKotlinModels` ?

Reply:
The main issue is https://github.com/apollographql/apollo-android/pull/1353. Once we add an empty source set to `compileJava` it will trigger an error... And I'm not sure if we can skip `compileJava` completely...

Reply:
I'm dumping this here while I'm thinking about it: We could have one artifact with 2 pluginIds. One would generate only java and the other only kotlin. Not sure what are the repercussions of that but that could be an option.